pkg/status/status.go

```diff
@@ -0,0 +1,155 @@
+	// that we can continue to work with Kubernetes clusters that don't have
+	// this API.
+	_, err := client.ClusterOperators().Get(clusterOperatorName, v1.GetOptions{})
+	coAPINotPresent := false
+	if meta.IsNoMatchError(err) {
+		log.Warning("ClusterOperator API not present")
+		coAPINotPresent = true
+	}
+
+	return &status{
+		client:          client,
+		coAPINotPresent: coAPINotPresent,
+		namespace:       namespace,
+	}
+}
+
+// SetFailing reports that operator has failed along with the error message
+func (s *status) SetFailing(message string) {
+	s.setStatus(configv1.OperatorFailing, message)
+}
+
+// SetFailing reports that the operator is available to process events
+func (s *status) SetAvailable(message string) {
+	s.setStatus(configv1.OperatorAvailable, message)
+}
+
+// ensureClusterOperator ensures that a ClusterOperator CR is present on the
+// cluster
+func (s *status) ensureClusterOperator() {
+	var err error
+	s.co, err = s.client.ClusterOperators().Get(clusterOperatorName, v1.GetOptions{})
+
+	if err == nil {
+		log.Info("Found existing ClusterOperator")
+		return
+	}
+
+	if err != nil && !errors.IsNotFound(err) {
+		log.Fatalf("Error %v getting ClusterOperator", err)
+	}
+
+	log.Info("Creating ClusterOperator")
+	s.co, err = s.client.ClusterOperators().Create(&configv1.ClusterOperator{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      clusterOperatorName,
+			Namespace: s.namespace,
+		},
+	})
+	if err != nil {
+		log.Fatalf("Error %v creating ClusterOperator", err)
+	}
+}
+
+// setStatus handles setting all the required fields for the given
+// ClusterStatusConditionType
+func (s *status) setStatus(condition configv1.ClusterStatusConditionType, message string) {
+	if s.coAPINotPresent {
+		return
+	}
+	log.Infof("Setting ClusterOperator condition: %s message: %s", condition, message)
+	s.ensureClusterOperator()
+	s.setStatusCondition(condition, message)
+	s.setOperandVersion()
+	s.updateStatus()
+}
+
+// setOperandVersion sets the operator version in the ClusterOperator Status
+func (s *status) setOperandVersion() {
+	// Report the operator version
+	operatorVersion := configv1.OperandVersion{
+		Name:    "operator",
+		Version: version.Version,
+	}
+	operatorhelpers.SetOperandVersion(&s.co.Status.Versions, operatorVersion)
+}
+
+// setStatusCondition sets the operator StatusCondition in the ClusterOperator Status
+func (s *status) setStatusCondition(condition configv1.ClusterStatusConditionType, message string) {
+	availableStatus := configv1.ConditionFalse
+	failingStatus := configv1.ConditionFalse
+	availableMessage := ""
+	failingMessage := ""
+
+	switch condition {
+	case configv1.OperatorAvailable:
+		availableStatus = configv1.ConditionTrue
+		availableMessage = message
+
+	case configv1.OperatorFailing:
+		failingStatus = configv1.ConditionTrue
+		failingMessage = message
+	}
+
+	time := v1.Now()
+	// As per the docs, all three StatusConditionTypes needs to be set with only
```

Comment:
This comment may be hard to understand for some. What docs are you referring to in this case? Can we link to them?

Reply:
Done. I have to say that even the docs don't clearly call this out. I found this out from a slack channel :-)